src/structure/utils.py

```diff
@@ -0,0 +1,385 @@
+# utils.py
+from __future__ import annotations
+import math
+from collections import defaultdict, namedtuple
+from typing import Dict, List, Tuple
+
+import numpy as np
+import biotite.structure as struc
+
+#________________HYDROGEN BONDS__________________________
+
+DA_MAX = 3.5           # Å donor–acceptor distance
+H_A_MAX = 2.6          # Å H–acceptor distance
+ANGLE_MIN = 120.0      # degrees minimum angle
+
+INCLUDE_WATER = False
+INCLUDE_LIGANDS = False
+
+DonorSite = namedtuple(
+    "DonorSite",
+    (
+        "res_key resname chain resi atom_name altloc "
+        "base_atom base_altloc coord_base coord_donor "
+        "has_explicit_H H_name H_altloc H_coord "
+        "is_backbone"
+    ),
+)
+
+AcceptorSite = namedtuple(
+    "AcceptorSite",
+    "res_key resname chain resi atom_name altloc coord is_backbone",
+)
+
+
+def _norm_vec(v: np.ndarray) -> np.ndarray:
+    x = np.linalg.norm(v)
+    return v / x if x > 1e-8 else v
+
+
+def angle_deg(u: np.ndarray, v: np.ndarray) -> float:
+    un = _norm_vec(u)
+    vn = _norm_vec(v)
+    dot = np.clip(np.dot(un, vn), -1.0, 1.0)
+    return float(math.degrees(math.acos(dot)))
+
+
+def _res_key(chain, resi, resname) -> str:
+    return f"{chain}:{int(resi)}:{resname}"
+
+
+def is_backbone_atom(atom_name: str) -> bool:
+    return atom_name in ("N", "CA", "C", "O", "OXT", "H", "H1", "H2", "H3")
+
+
+def norm_alt(a) -> str:
+    if a is None:
+        return ""
+    s = str(a).strip()
+    return s.upper() if s else ""
+
+
+def altloc_compatible(d_alt, a_alt) -> bool:
+    d = norm_alt(d_alt)
+    a = norm_alt(a_alt)
+    if d == "" or a == "":
+        return True
+    return d == a
+
+def _split_by_residue(arr: struc.AtomArray):
+    """
+    Yield per residue:
+      (resname, chain_id, res_id, idxs, base_arr)
+
+    Residues are sorted by (chain_id, res_id).
+    """
+    if arr.array_length() == 0:
+        return
+
+    order = np.lexsort((arr.res_id, arr.chain_id))
+    arr = arr[order]
+
+    starts = [0]
+    for i in range(1, arr.array_length()):
+        if (arr.chain_id[i] != arr.chain_id[i - 1] or
+                arr.res_id[i] != arr.res_id[i - 1]):
+            starts.append(i)
+    starts.append(arr.array_length())
```

Comment:
Isn't this the same as `get_residue_starts`? Is it important that np.lexsort() is used as the order rather than default order in the AtomArray?